embrio-async/src/lib.rs

```diff
@@ -77,7 +96,7 @@ enum FutureImplState<C, G> {
 // `FutureImplState::NotStarted(C)` to `FutureImplState::Started(G)` and the generator executes
 // till the first yield point.
 struct FutureImpl<C, G> {
-    context: *mut task::Context<'static>,
+    context: MaybeUninit<*mut task::Context<'static>>,
```

Comment:
What's the advantage of using a `MaybeUninit` here?

(I could see switching to `MaybeUninit<NonNull<task::Context<'static>>>` being potentially useful to provide a niche if this is included in an enum, but I don't see what advantage an uninit raw pointer brings).

Reply:
(Actually, with `NonNull` there's also `NonNull::dangling` as an uninit sentinel, so that wouldn't need `MaybeUninit` either)

Reply:
> What's the advantage of using a MaybeUninit here?

This is just me being picky and partially why I separated that change into its own commit, but... when constructing the future there's just no requirement for it to be initialized, so the `ptr::null_mut()` initializer is a dead store. I imagine that gets recognized/eliminated in cases where the future is inlined or immediately polled, but a lot of the point of these changes is to help guide codegen in cases where that doesn't or can't happen.

A single store of 0 is what, 2~4 bytes at most, which is pretty small in comparison to the impact of the other changes, but it's still potentially something!

> I could see switching to MaybeUninit<NonNull<task::Context<'static>>> being potentially useful to provide a niche if this is included in an enum, but I don't see what advantage an uninit raw pointer brings

Oh huh I actually had no idea `NonNull` had been stabilized for some reason (not that that should discourage me from using it in this codebase I imagine), that is better. Yeah, in some of my other changes I've moved the `context` field directly into the `Started` state enum variant alongside `G` rather than placing it in the parent struct, but I left that out to minimize diff noise and unnecessary shuffling (though introducing `LateInit<G>` made me reconsider whether I should've just done it anyway)... That's also why I had the `Started` variant contain the `MaybeUninit<G>`, because the enum discriminant needed to be switched over before getting the pointer to the context field and calling the constructor with it.

> there's also NonNull::dangling as an uninit sentinel, so that wouldn't need MaybeUninit either

Hm, that is cleaner than just throwing `null_mut()` in there though does still introduce the dead store I was aiming to avoid.